File: python/mops/src/mops/sap.py
import numpy as np

from . import _dispatch
from .checks import _check_sap, _check_sap_vjp, _check_sap_vjp_vjp
from .dispatch_operation import dispatch_operation
from .utils import mops_tensor, null_mops_tensor_like


def sparse_accumulation_of_products(
    A, B, C, indices_A, indices_B, indices_output, output_size
):
    A = _dispatch.make_contiguous(A)
    B = _dispatch.make_contiguous(B)
    C = _dispatch.make_contiguous(C)
    indices_A = _dispatch.make_contiguous(indices_A)
    indices_B = _dispatch.make_contiguous(indices_B)
    indices_output = _dispatch.make_contiguous(indices_output)

    _check_sap(A, B, C, indices_A, indices_B, indices_output, output_size)

    indices_A = indices_A.astype(np.int32)
    indices_B = indices_B.astype(np.int32)
    indices_output = indices_output.astype(np.int32)

    output = _dispatch.empty_like((A.shape[0], output_size), A)

    function = dispatch_operation(
        "sparse_accumulation_of_products",
        A,
    )

    function(
        mops_tensor(output),
        mops_tensor(A),
        mops_tensor(B),
        mops_tensor(C),
        mops_tensor(indices_A),
        mops_tensor(indices_B),
        mops_tensor(indices_output),
    )

    return output


def sparse_accumulation_of_products_vjp(
    grad_output,
    A,
    B,
    C,
    indices_A,
    indices_B,
    indices_output,
    compute_grad_A=True,
    compute_grad_B=True,
):
    _check_sap_vjp(
        grad_output, A, B, C, indices_A, indices_B, indices_output, grad_output.shape[1]
    )

    grad_output = _dispatch.make_contiguous(grad_output)
    A = _dispatch.make_contiguous(A)
    B = _dispatch.make_contiguous(B)
    C = _dispatch.make_contiguous(C)
    indices_A = _dispatch.make_contiguous(indices_A)
    indices_B = _dispatch.make_contiguous(indices_B)
    indices_output = _dispatch.make_contiguous(indices_output)

    indices_A = indices_A.astype(np.int32)
    indices_B = indices_B.astype(np.int32)
    indices_output = indices_output.astype(np.int32)

    if compute_grad_A:
        grad_A = _dispatch.empty_like(A.shape, A)
        mops_grad_A = mops_tensor(grad_A)
    else:
        grad_A = None
        mops_grad_A = null_mops_tensor_like(A)

    if compute_grad_B:
        grad_B = _dispatch.empty_like(B.shape, B)
        mops_grad_B = mops_tensor(grad_B)
    else:
        grad_B = None
        mops_grad_B = null_mops_tensor_like(B)

    function = dispatch_operation(
        "sparse_accumulation_of_products_vjp",
        A,
    )

    function(
        mops_grad_A,
        mops_grad_B,
        mops_tensor(grad_output),
        mops_tensor(A),
        mops_tensor(B),
        mops_tensor(C),
        mops_tensor(indices_A),
        mops_tensor(indices_B),
        mops_tensor(indices_output),
    )

    return grad_A, grad_B


def sparse_accumulation_of_products_vjp_vjp(
    grad_grad_A,
    grad_grad_B,
    grad_output,
    A,
    B,
    C,
    indices_A,
    indices_B,
    indices_output,
    compute_grad_grad_output=True,
    compute_grad_A_2=True,
    compute_grad_B_2=True,
):
    _check_sap_vjp_vjp(
        grad_grad_A,
        grad_grad_B,
        grad_output,
        A,
        B,
        C,
        indices_A,
        indices_B,
        indices_output,
        grad_output.shape[1],
    )

    grad_grad_A = _dispatch.make_contiguous(grad_grad_A)
    grad_grad_B = _dispatch.make_contiguous(grad_grad_B)
    A = _dispatch.make_contiguous(A)
    B = _dispatch.make_contiguous(B)
    C = _dispatch.make_contiguous(C)
    indices_A = _dispatch.make_contiguous(indices_A)
    indices_B = _dispatch.make_contiguous(indices_B)
    indices_output = _dispatch.make_contiguous(indices_output)

    indices_A = indices_A.astype(np.int32)
    indices_B = indices_B.astype(np.int32)
    indices_output = indices_output.astype(np.int32)

    if compute_grad_grad_output:
        grad_grad_output = _dispatch.empty_like(
            (grad_grad_A.shape[1], A.shape[1], B.shape[1]), grad_grad_A
        )
        mops_grad_grad_output = mops_tensor(grad_grad_output)
    else:
        grad_grad_output = None
        mops_grad_grad_output = null_mops_tensor_like(grad_grad_A)

    if compute_grad_A_2:
        grad_A_2 = _dispatch.empty_like(A.shape, A)
        mops_grad_A_2 = mops_tensor(grad_A_2)
    else:
        grad_A_2 = None
        mops_grad_A_2 = null_mops_tensor_like(A)

    if compute_grad_B_2:
        grad_B_2 = _dispatch.empty_like(B.shape, B)
        mops_grad_B_2 = mops_tensor(grad_B_2)
    else:
        grad_B_2 = None
        mops_grad_B_2 = null_mops_tensor_like(B)

    function = dispatch_operation(
        "sparse_accumulation_of_products_vjp_vjp",
        A,
    )

    function(
        mops_grad_grad_output,
        mops_grad_A_2,
        mops_grad_B_2,
        (
            mops_tensor(grad_grad_A)
            if grad_grad_A is not None
            else null_mops_tensor_like(A)
        ),
        (
            mops_tensor(grad_grad_B)
            if grad_grad_B is not None
            else null_mops_tensor_like(B)
        ),
        mops_tensor(grad_output),
        mops_tensor(A),
        mops_tensor(B),
        mops_tensor(C),
        mops_tensor(indices_A),
        mops_tensor(indices_B),
        mops_tensor(indices_output),
    )

    return grad_grad_output, grad_A_2, grad_B_2

File: mops-torch/src/sap.cpp
#include "mops/torch/sap.hpp"
#include "mops/torch/utils.hpp"

using namespace mops_torch;

torch::Tensor mops_torch::sparse_accumulation_of_products(
    torch::Tensor A,
    torch::Tensor B,
    torch::Tensor C,
    torch::Tensor indices_A,
    torch::Tensor indices_B,
    torch::Tensor indices_output,
    int64_t output_size
) {
    return SparseAccumulationOfProducts::apply(
        A, B, C, indices_A, indices_B, indices_output, output_size
    );
}

torch::Tensor SparseAccumulationOfProducts::forward(
    torch::autograd::AutogradContext* ctx,
    torch::Tensor A,
    torch::Tensor B,
    torch::Tensor C,
    torch::Tensor indices_A,
    torch::Tensor indices_B,
    torch::Tensor indices_output,
    int64_t output_size
) {
    details::check_all_same_device({A, B, C, indices_A, indices_B, indices_output});
    details::check_floating_dtype({A, B, C});
    details::check_integer_dtype({indices_A, indices_B, indices_output});
    details::check_number_of_dimensions(A, 2, "A", "sparse_accumulation_of_products");
    details::check_number_of_dimensions(B, 2, "B", "sparse_accumulation_of_products");
    details::check_number_of_dimensions(indices_A, 1, "indices_A", "sparse_accumulation_of_products");
    details::check_number_of_dimensions(indices_B, 1, "indices_B", "sparse_accumulation_of_products");
    details::check_number_of_dimensions(
        indices_output, 1, "indices_output", "sparse_accumulation_of_products"
    );
    // Shape consistency checks are performed inside
    // mops::sparse_accumulation_of_products

    torch::Tensor output;
    if (A.device().is_cpu()) {
        output = torch::empty(
            {A.size(0), output_size},
            torch::TensorOptions().dtype(A.scalar_type()).device(A.device())
        );

        AT_DISPATCH_FLOATING_TYPES(A.scalar_type(), "sparse_accumulation_of_products", [&]() {
            mops::sparse_accumulation_of_products<scalar_t>(
                details::torch_to_mops_2d<scalar_t>(output),
                details::torch_to_mops_2d<scalar_t>(A),
                details::torch_to_mops_2d<scalar_t>(B),
                details::torch_to_mops_1d<scalar_t>(C),
                details::torch_to_mops_1d<int32_t>(indices_A),
                details::torch_to_mops_1d<int32_t>(indices_B),
                details::torch_to_mops_1d<int32_t>(indices_output)
            );
        });
    } else if (A.device().is_cuda()) {
        output = torch::empty(
            {A.size(0), output_size},
            torch::TensorOptions().dtype(A.scalar_type()).device(A.device())
        );

        AT_DISPATCH_FLOATING_TYPES(A.scalar_type(), "sparse_accumulation_of_products", [&]() {
            mops::cuda::sparse_accumulation_of_products<scalar_t>(
                details::torch_to_mops_2d<scalar_t>(output),
                details::torch_to_mops_2d<scalar_t>(A),
                details::torch_to_mops_2d<scalar_t>(B),
                details::torch_to_mops_1d<scalar_t>(C),
                details::torch_to_mops_1d<int32_t>(indices_A),
                details::torch_to_mops_1d<int32_t>(indices_B),
                details::torch_to_mops_1d<int32_t>(indices_output)
            );
        });
    } else {
        C10_THROW_ERROR(
            ValueError,
            "sparse_accumulation_of_products is not implemented for device " + A.device().str()
        );
    }

    if (A.requires_grad() || B.requires_grad()) {
        ctx->save_for_backward({A, B, C, indices_A, indices_B, indices_output});
    }

    return {output};
}

std::vector<torch::Tensor> SparseAccumulationOfProducts::backward(
    torch::autograd::AutogradContext* ctx, std::vector<torch::Tensor> grad_outputs
) {
    auto saved_variables = ctx->get_saved_variables();
    auto A = saved_variables[0];
    auto B = saved_variables[1];
    auto C = saved_variables[2];
    auto indices_A = saved_variables[3];
    auto indices_B = saved_variables[4];
    auto indices_output = saved_variables[5];

    auto grad_output = grad_outputs[0].contiguous();

    auto results = SparseAccumulationOfProductsBackward::apply(
        grad_output, A, B, C, indices_A, indices_B, indices_output
    );
    auto grad_A = results[0];
    if (grad_output.requires_grad() || A.requires_grad() || B.requires_grad()) {
        ctx->save_for_backward({grad_output, A, B, C, indices_A, indices_B, indices_output});
    }
    auto grad_B = results[1];

    return {
        grad_A,
        grad_B,
        torch::Tensor(),
        torch::Tensor(),
        torch::Tensor(),
        torch::Tensor(),
        torch::Tensor()
    };
}

std::vector<torch::Tensor> SparseAccumulationOfProductsBackward::forward(
    torch::autograd::AutogradContext* ctx,
    torch::Tensor grad_output,
    torch::Tensor A,
    torch::Tensor B,
    torch::Tensor C,
    torch::Tensor indices_A,
    torch::Tensor indices_B,
    torch::Tensor indices_output
) {
    if (C.requires_grad()) {
        C10_THROW_ERROR(
            ValueError,
            "gradients with respect to C are not supported "
            "in sparse_accumulation_of_products"
        );
    }

    auto grad_A = torch::Tensor();
    auto grad_B = torch::Tensor();

    if (A.device().is_cpu()) {
        AT_DISPATCH_FLOATING_TYPES(A.scalar_type(), "sparse_accumulation_of_products_vjp", [&]() {
            auto mops_grad_A = mops::Tensor<scalar_t, 2>{nullptr, {0, 0}};
            if (A.requires_grad()) {
                grad_A = torch::empty_like(A);
                mops_grad_A = details::torch_to_mops_2d<scalar_t>(grad_A);
            }

            auto mops_grad_B = mops::Tensor<scalar_t, 2>{nullptr, {0, 0}};
            if (B.requires_grad()) {
                grad_B = torch::empty_like(B);
                mops_grad_B = details::torch_to_mops_2d<scalar_t>(grad_B);
            }

            mops::sparse_accumulation_of_products_vjp<scalar_t>(
                mops_grad_A,
                mops_grad_B,
                details::torch_to_mops_2d<scalar_t>(grad_output),
                details::torch_to_mops_2d<scalar_t>(A),
                details::torch_to_mops_2d<scalar_t>(B),
                details::torch_to_mops_1d<scalar_t>(C),
                details::torch_to_mops_1d<int32_t>(indices_A),
                details::torch_to_mops_1d<int32_t>(indices_B),
                details::torch_to_mops_1d<int32_t>(indices_output)
            );
        });
    } else if (A.device().is_cuda()) {
        AT_DISPATCH_FLOATING_TYPES(A.scalar_type(), "sparse_accumulation_of_products_vjp", [&]() {
            auto mops_grad_A = mops::Tensor<scalar_t, 2>{nullptr, {0, 0}};
            if (A.requires_grad()) {
                grad_A = torch::empty_like(A);
                mops_grad_A = details::torch_to_mops_2d<scalar_t>(grad_A);
            }

            auto mops_grad_B = mops::Tensor<scalar_t, 2>{nullptr, {0, 0}};
            if (B.requires_grad()) {
                grad_B = torch::empty_like(B);
                mops_grad_B = details::torch_to_mops_2d<scalar_t>(grad_B);
            }

            mops::cuda::sparse_accumulation_of_products_vjp<scalar_t>(
                mops_grad_A,
                mops_grad_B,
                details::torch_to_mops_2d<scalar_t>(grad_output),
                details::torch_to_mops_2d<scalar_t>(A),
                details::torch_to_mops_2d<scalar_t>(B),
                details::torch_to_mops_1d<scalar_t>(C),
                details::torch_to_mops_1d<int32_t>(indices_A),
                details::torch_to_mops_1d<int32_t>(indices_B),
                details::torch_to_mops_1d<int32_t>(indices_output)
            );
        });
    } else {
        C10_THROW_ERROR(
            ValueError,
            "sparse_accumulation_of_products is not implemented for device " + A.device().str()
        );
    }

    if (grad_output.requires_grad() || A.requires_grad() || B.requires_grad()) {
        ctx->save_for_backward({grad_output, A, B, C, indices_A, indices_B, indices_output});
    }

    return {grad_A, grad_B};
}

std::vector<torch::Tensor> SparseAccumulationOfProductsBackward::backward(
    torch::autograd::AutogradContext* ctx, std::vector<torch::Tensor> grad_outputs
) {
    auto saved_variables = ctx->get_saved_variables();
    auto grad_output = saved_variables[0];
    auto A = saved_variables[1];
    auto B = saved_variables[2];
    auto C = saved_variables[3];
    auto indices_A = saved_variables[4];
    auto indices_B = saved_variables[5];
    auto indices_output = saved_variables[6];
    auto grad_grad_A = grad_outputs[0].contiguous();
    auto grad_grad_B = grad_outputs[1].contiguous();

    auto grad_grad_output = torch::Tensor();
    auto grad_A_2 = torch::Tensor();
    auto grad_B_2 = torch::Tensor();

    if (A.device().is_cpu()) {
        AT_DISPATCH_FLOATING_TYPES(A.scalar_type(), "sparse_accumulation_of_products_vjp_vjp", [&]() {
            auto mops_grad_grad_output = mops::Tensor<scalar_t, 2>{nullptr, {0, 0}};
            if (grad_output.requires_grad()) {
                grad_grad_output = torch::empty_like(grad_output);
                mops_grad_grad_output = details::torch_to_mops_2d<scalar_t>(grad_grad_output);
            }

            auto mops_grad_A_2 = mops::Tensor<scalar_t, 2>{nullptr, {0, 0}};
            if (A.requires_grad()) {
                grad_A_2 = torch::empty_like(A);
                mops_grad_A_2 = details::torch_to_mops_2d<scalar_t>(grad_A_2);
            }

            auto mops_grad_B_2 = mops::Tensor<scalar_t, 2>{nullptr, {0, 0}};
            if (B.requires_grad()) {
                grad_B_2 = torch::empty_like(B);
                mops_grad_B_2 = details::torch_to_mops_2d<scalar_t>(grad_B_2);
            }

            auto mops_grad_grad_A = mops::Tensor<scalar_t, 2>{nullptr, {0, 0}};
            if (grad_grad_A.defined()) {
                mops_grad_grad_A = details::torch_to_mops_2d<scalar_t>(grad_grad_A);
            }

            auto mops_grad_grad_B = mops::Tensor<scalar_t, 2>{nullptr, {0, 0}};
            if (grad_grad_B.defined()) {
                mops_grad_grad_B = details::torch_to_mops_2d<scalar_t>(grad_grad_B);
            }

            mops::sparse_accumulation_of_products_vjp_vjp<scalar_t>(
                mops_grad_grad_output,
                mops_grad_A_2,
                mops_grad_B_2,
                mops_grad_grad_A,
                mops_grad_grad_B,
                details::torch_to_mops_2d<scalar_t>(grad_output),
                details::torch_to_mops_2d<scalar_t>(A),
                details::torch_to_mops_2d<scalar_t>(B),
                details::torch_to_mops_1d<scalar_t>(C),
                details::torch_to_mops_1d<int32_t>(indices_A),
                details::torch_to_mops_1d<int32_t>(indices_B),
                details::torch_to_mops_1d<int32_t>(indices_output)
            );
        });
    } else if (A.device().is_cuda()) {
        C10_THROW_ERROR(
            ValueError, "sparse_accumulation_of_products_vjp_vjp is not implemented for CUDA yet"
        );
    } else {
        C10_THROW_ERROR(
            ValueError,
            "sparse_accumulation_of_products_vjp_vjp is not implemented for device " +
                A.device().str()
        );
    }

    return {
        grad_grad_output,
        grad_A_2,
        grad_B_2,
        torch::Tensor(),
        torch::Tensor(),
        torch::Tensor(),
        torch::Tensor()
    };
}
